### PR TITLE
Make sure not to leak intermediate result folders on the workers

### LINKED ISSUE #2420: Citus fails to clear intermediate results when 2PC involved (e.g., pull-push execution with modifications)
While investigating #2418, I hit another issue.

One interesting thing about this is that when 2PC involved, Citus leaks intermediate results on the worker nodes. I don't think that the issue is directly related with the following, but wanted to share some of my observations:

- Intermediate result folder name relies on distributed transaction id
- We call `UnSetDistributedTransactionId()` on  `XACT_EVENT_PREPARE`
- But, we call `RemoveIntermediateResultsDirectory()` on `XACT_EVENT_PRE_COMMIT`, which happens after `XACT_EVENT_PREPARE`
- So, `ls ` on the workers `/Users/onderkalaci/Documents/data_dir/worker_9700/base/pgsql_job_cache` could show the intermediate results that are not removed:
```
 ls
10_0_1		10_0_120	10_0_140	10_0_16		10_0_179	10_0_198	10_0_216	10_0_235	10_0_254	10_0_41		10_0_61		10_0_82
10_0_100	10_0_121	10_0_141	10_0_160	10_0_18		10_0_199	10_0_217	10_0_236	10_0_255	10_0_42		10_0_62		10_0_83
10_0_101	10_0_122	10_0_142	10_0_161	10_0_180	10_0_2		10_0_218	10_0_237	10_0_256	10_0_43		10_0_63		10_0_84
10_0_102	10_0_123	10_0_143	10_0_162	10_0_181	10_0_20		10_0_219	10_0_238	10_0_257	10_0_44		10_0_64		10_0_85
10_0_103	10_0_124	10_0_144	10_0_163	10_0_182	10_0_200	10_0_22		10_0_239	10_0_26		10_0_45		10_0_65		10_0_86
10_0_104	10_0_125	10_0_145	10_0_164	10_0_183	10_0_201	10_0_220	10_0_24		10_0_27		10_0_46		10_0_66		10_0_87
10_0_105	10_0_126	10_0_146	10_0_165	10_0_184	10_0_202	10_0_221	10_0_240	10_0_28		10_0_47		10_0_67		10_0_88
10_0_106	10_0_127	10_0_147	10_0_166	10_0_185	10_0_203	10_0_222	10_0_241	10_0_29		10_0_48		10_0_68		10_0_89
10_0_107	10_0_128	10_0_148	10_0_167	10_0_186	10_0_204	10_0_223	10_0_242	10_0_3		10_0_49		10_0_69		10_0_9
10_0_108	10_0_129	10_0_149	10_0_168	10_0_187	10_0_205	10_0_224	10_0_243	10_0_30		10_0_5		10_0_70		10_0_90
10_0_109	10_0_130	10_0_15		10_0_169	10_0_188	10_0_206	10_0_225	10_0_244	10_0_31		10_0_50		10_0_71		10_0_91
10_0_110	10_0_131	10_0_150	10_0_17		10_0_189	10_0_207	10_0_226	10_0_245	10_0_32		10_0_51		10_0_72		10_0_92
10_0_111	10_0_132	10_0_151	10_0_170	10_0_19		10_0_208	10_0_227	10_0_246	10_0_33		10_0_52		10_0_73		10_0_93
10_0_112	10_0_133	10_0_152	10_0_171	10_0_190	10_0_209	10_0_228	10_0_247	10_0_34		10_0_53		10_0_74		10_0_94
10_0_113	10_0_134	10_0_153	10_0_172	10_0_191	10_0_21		10_0_229	10_0_248	10_0_35		10_0_54		10_0_75		10_0_95
10_0_114	10_0_135	10_0_154	10_0_173	10_0_192	10_0_210	10_0_23		10_0_249	10_0_36		10_0_55		10_0_76		10_0_96
10_0_115	10_0_136	10_0_155	10_0_174	10_0_193	10_0_211	10_0_230	10_0_25		10_0_37		10_0_56		10_0_77		10_0_98
10_0_116	10_0_137	10_0_156	10_0_175	10_0_194	10_0_212	10_0_231	10_0_250	10_0_38		10_0_57		10_0_78		10_0_99
10_0_117	10_0_138	10_0_157	10_0_176	10_0_195	10_0_213	10_0_232	10_0_251	10_0_39		10_0_58		10_0_79
10_0_118	10_0_139	10_0_158	10_0_177	10_0_196	10_0_214	10_0_233	10_0_252	10_0_4		10_0_59		10_0_80
10_0_119	10_0_14		10_0_159	10_0_178	10_0_197	10_0_215	10_0_234	10_0_253	10_0_40		10_0_60		10_0_81

```

- If you log the folder create/drops, you'd see something like the following, which is clearly the behaviour defined on `IntermediateResultsDirectory()`:
```
2018-10-01 12:37:26.398 EEST [66342] LOG:  resultsDirectory is created :base/pgsql_job_cache/10_0_21
2018-10-01 12:37:26.863 EEST [66335] LOG:  resultsDirectory is removed:base/pgsql_job_cache/10_66335
```

This is cleared with a restart.

### PR DESCRIPTION
**Pull request by onderkalaci**
Fixes #2420 

To ensure that Citus never leaks any intermediate results (e.g., recursive planning, real-time or task-tracker), I added the test at the end of the regression test suite.

But it seems like we've some more edge cases in real-time (or task-tracker?) because the tests sometimes fail with the following output indicating that we've one folder leaking at the end of the regression tests:
```SQL
 SELECT run_command_on_workers($$SELECT pg_ls_dir('base/pgsql_job_cache')$$);
  run_command_on_workers 
-------------------------
+--------------------------------------
  (localhost,57637,t,"")
- (localhost,57638,t,"")
+ (localhost,57638,t,job_238202912770)
 (2 rows)
```

And, that seems to be caused by not handling `task-tracker` failures well, because if I execute `multi_task_string_size.sql` multiple times, I'd see one intermediate result per call.

So, I could overlook the task-tracker failures and only test recursive planning. 

In any case, I'm still investigating the task-tracker issue some more.

### CODE REVIEW

**Comment by onderkalaci** [comment]
> But it seems like we've some more edge cases in real-time (or task-tracker?) because the tests sometimes fail with the following output indicating that we've one folder leaking at the end of the regression tests:

And note that this happens irrespective of the code changes in the PR. Even if I revert the code changes, the test still fails.

**Comment by onderkalaci** [comment]
I prefer to not handle task-tracker issue on this PR. That only happens on failures and we already have some edge cases with task-tracker.  Once we merge this, I'll open an issue for it.

File: src/backend/distributed/transaction/transaction_management.c [hunk 1]
```diff
@@ -281,12 +281,28 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
...
 		case XACT_EVENT_PREPARE:
 		{
+			/*
+			 * We should remove the intermediate results before unsetting the
+			 * distributed transaction id. That is necessary, otherwise Citus
+			 * would try to remove a non-existing folder and leak some of the
+			 * existing folders that are associated with distributed transaction
+			 * ids on the worker nodes.
+			 */
+			RemoveIntermediateResultsDirectory();
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_management.c]: Worth adding a note that this is called only on the worker, whereas most code in this function addresses the coordinator.

File: src/test/regress/sql/ensure_no_intermediate_data_leak.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+------
+-- THIS TEST SHOULD BE EXECUTED AT THE END OF THE REGRESSION TEST SUIT
```
Comment (marcocitus) [inline src/test/regress/sql/ensure_no_intermediate_data_leak.sql]: suit -> suite

File: src/test/regress/expected/ensure_no_intermediate_data_leak.out [hunk 3]
```diff
@@ -0,0 +1,17 @@
+------
+-- THIS TEST SHOULD BE EXECUTED AT THE END OF THE REGRESSION TEST SUIT
+-- TO MAKE SURE THAT WE CLEAR ALL INTERMEDIATE RESULTS
+-- EITHER ON THE COORDINATOR OR ON THE WORKERS
```
Comment (marcocitus) [inline src/test/regress/expected/ensure_no_intermediate_data_leak.out]: Either on .. or -> on both .. and

File: src/test/regress/sql/dml_recursive.sql [hunk 4]
```diff
@@ -261,5 +261,9 @@ FROM
 WHERE 
 	distributed_table.tenant_id = local_table.id;
 
+WITH dirs AS (SELECT pg_ls_dir('base/pgsql_job_cache') as dir) SELECT array_agg(dir) FROM dirs;
+SELECT run_command_on_workers($$WITH dirs AS (SELECT pg_ls_dir('base/pgsql_job_cache') as dir) SELECT array_agg(dir) FROM dirs;$$);
```
Comment (marcocitus) [inline src/test/regress/sql/dml_recursive.sql]: missing output file, probably

**Comment by onderkalaci** [comment]
> missing output file, probably
Oh no, I realized that appveyor fails, so I wanted to see what kind of output I'd get.

```
+SELECT run_command_on_workers($$WITH dirs AS (SELECT pg_ls_dir('base/pgsql_job_cache') as dir) SELECT array_agg(dir) FROM dirs;$$);
+                                              run_command_on_workers                                              
+------------------------------------------------------------------------------------------------------------------
+ (localhost,57637,t,"{10_0_1001,10_0_1029,10_0_1032,10_0_1036,10_0_1037,10_0_1039,10_0_1040,10_0_1744,10_0_994}")
+ (localhost,57638,t,"{10_0_1001,10_0_1029,10_0_1032,10_0_1036,10_0_1037,10_0_1039,10_0_1040,10_0_1744,10_0_994}")
 (2 rows)
 
```

So, I'm planning to do the following:
- Change the test such that the same call is only done at the end of `dml_recursive.sql`
- Open an issue that we've issues with task-tracker failures
- Open another issue that we've some issues with windows/appveyor

@marcocitus does that sound OK?

**Comment by marcocitus** [comment]
hmz, so are we not removing intermediate result directories at all on Windows?

**Comment by onderkalaci** [comment]
> hmz, so are we not removing intermediate result directories at all on Windows?

I don't think we always skip removing, otherwise, I'd expect a lot more folders. It might be possible that we're not removing in case of failures or maybe some other edges?

**Comment by onderkalaci** [comment]
> windows trouble

Appveyor is even worse 😄

File: src/test/regress/expected/ensure_no_intermediate_data_leak.out [hunk 5]
```diff
@@ -0,0 +1,17 @@
+------
+-- THIS TEST SHOULD BE EXECUTED AT THE END OF THE REGRESSION TEST SUITE
```
Comment (onderkalaci) [inline src/test/regress/expected/ensure_no_intermediate_data_leak.out]: I'll update this comment. 

@marcocitus do you have any other comments before merging this PR?

I'll open an issue describing the cases where Citus fails to clear all intermediate results:
- Failures in task-tracker
- Windows (not sure about the source)

**Comment by metdos** [comment]
Do we need a description for this?

**Comment by onderkalaci** [comment]
> Do we need a description for this?

Yes, I forgot to add. Now added (cc @velioglu)
